File: 0199-binary-tree-right-side-view/0199-binary-tree-right-side-view.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> rightSideView(TreeNode* root) {
        vector<int> res;
        helper(res, 0, root);
        return res;
    }
    
    void helper(vector<int>& res, int level, TreeNode* root){
        if (!root){return;}
        if (res.size() == level){
            res.push_back(root->val);
        }
        helper(res, level+1, root->right);
        helper(res, level+1, root->left);
    }
};

File: 0199-binary-tree-right-side-view/0199-binary-tree-right-side-view.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def __init__(self):
        self.result = []
        self.levels = dict()

    def rightSideView(self, root):
        if not root:
            return []
        self.do(root, 0)
        return self.result
        
    def do(self, node, level):
        if not node:
            return
        
        if level not in self.levels:
            self.levels[level] = True
            self.result.append(node.val)

        self.do(node.right, level+1)
        self.do(node.left, level+1)
        return
        
